### PR TITLE
Testing for EKS support in kubeprod

### PR DESCRIPTION
**Pull request by falfaro**

### CODE REVIEW

**Comment by anguslees** [comment]
>  I am blindly adding manifests/tests/eks.jsonnet [..] but I believe this file is unused?

It is used by `make -C manifests validate`, run (early) by Jenkinsfile.

File: Jenkinsfile [hunk 1]
```diff
@@ -488,6 +499,91 @@ gcloud container clusters create ${clusterName} \
...
+                                            writeFile([file: "${clusterName}-autogen.json", text: """
+{
+  "dnsZone": "${dnsZone}",
+  "contactEmail": "${adminEmail}",
+  "externalDns": {
+    "aws_access_key_id": "${AWS_ACCESS_KEY_ID}",
+    "aws_secret_access_key": "${AWS_SECRET_ACCESS_KEY}"
+  },
+  "oauthProxy": {
+    "client_id": "${fixme}",
```
Comment (anguslees) [inline Jenkinsfile]: These need appropriate values

File: Jenkinsfile [hunk 2]
```diff
@@ -488,6 +499,91 @@ gcloud container clusters create ${clusterName} \
...
+                                            writeFile([file: 'kubeprod-manifest.jsonnet', text: """
+(import "manifests/platforms/eks.jsonnet") {
+  config:: import "${clusterName}-autogen.json",
+  letsencrypt_environment: "staging",
+  prometheus+: import "tests/testdata/prometheus-crashloop-alerts.jsonnet",
+}
+"""
+                                            ])
+                                        }
+                                    }{
```
Comment (anguslees) [inline Jenkinsfile]: Syntax error here somewhere:
```console
org.codehaus.groovy.control.MultipleCompilationErrorsException: startup failed:
WorkflowScript: 567: unexpected token: { @ line 567, column 38.
                                       }{
                                        ^
```

File: Jenkinsfile [hunk 3]
```diff
@@ -488,6 +499,91 @@ gcloud container clusters create ${clusterName} \
...
+                                        container('aws') {
+                                            def output = sh(returnStdout: true, script: """
+id=$(aws route53 list-hosted-zones --query 'HostedZones[?Name==`bkpr.run.`].Id' --output text)
+aws route53 get-hosted-zone --id $id --query DelegationSet.NameServers
+""")
+                                            def nameServers = readJson(text: output)
+                                            insertGlueRecords(dnsPrefix, nameServers, "60", parentZone, parentZoneResourceGroup)
+                                        }
+                                    }
+                                }
```
Comment (anguslees) [inline Jenkinsfile]: oops, looks like I forgot to add cleanup code in a `finally` unwind about here.  We should call `eksctl delete cluster ...`

File: Jenkinsfile [hunk 4]
```diff
@@ -488,6 +499,91 @@ gcloud container clusters create ${clusterName} \
...
     }
 
+    def eksVersions = ["1.10", "1.11"]
+    for (x in eksVersions) {
+        def kversion = x // local bind required because closures
+        def platform = "eks-${kversion}"
+        platforms[platform] = {
+            stage(platform) {
+                node(label) {
+                    withCredentials([[$class: 'AmazonWebServicesCredentialsBinding', credentialsId: 'aws-kubeprod-jenkins']]) {
```
Comment (anguslees) [inline Jenkinsfile]: This jenkins secret still needs to be created.

File: Jenkinsfile (hunk 3 above)
Comment (falfaro) [reply Jenkinsfile]: Done.

File: Jenkinsfile (hunk 1 above)
Comment (falfaro) [reply Jenkinsfile]: Fixed.

File: Jenkinsfile (hunk 3 above)
Comment (falfaro) [reply Jenkinsfile]: Fixed.

File: Jenkinsfile (hunk 4 above)
Comment (falfaro) [reply Jenkinsfile]: Added to Jenkins (and changed the secret ID in Jenkins to a more meaningful name)

**Comment by falfaro** [comment]
@anguslees PTAL

File: Jenkinsfile [hunk 5]
```diff
@@ -23,6 +24,8 @@ def withGo(Closure body) {
             "GOPATH+WS=${env.WORKSPACE}",
             "PATH+GOBIN=${env.WORKSPACE}/bin",
             "HOME=${env.WORKSPACE}",
+            // https://docs.aws.amazon.com/eks/latest/userguide/install-aws-iam-authenticator.html
+            "PATH+AWSIAMAUTHENTICATOR=${tool 'aws-iam-authenticator'}"
```
Comment (sameersbn) [inline Jenkinsfile]: I would suggest performing the tool installation only where it will be used. By defining it here, we're installing it every time the `withGo` function is called.

File: Jenkinsfile [hunk 6]
```diff
@@ -490,6 +506,107 @@ az aks create                      \
...
+            stage(platform) {
+                node(label) {
+                    withCredentials([[
+                        $class: 'AmazonWebServicesCredentialsBinding',
+                        credentialsId: 'aws-eks-kubeprod-jenkins',
+                        accessKeyVariable: 'AWS_ACCESS_KEY_ID',
+                        secretKeyVariable: 'AWS_SECRET_ACCESS_KEY',
+                    ]]) {
+                        withEnv([
+                            "AWS_DEFAULT_REGION=eu-central-1",
```
Comment (sameersbn) [inline Jenkinsfile]: FWIW, you can add this as an environment variable for the aws-cli container in the PodTemplate

File: Jenkinsfile [hunk 7]
```diff
@@ -155,6 +162,16 @@ spec:
     tty: true
     command:
     - 'cat'
+  - name: 'eksctl'
+    image: weaveworks/eksctl
```
Comment (sameersbn) [inline Jenkinsfile]: We should be specifying the docker image tag

File: Jenkinsfile [hunk 8]
```diff
@@ -155,6 +162,16 @@ spec:
...
+    tty: true
+    command:
+    - 'cat'
+  - name: 'aws'
+    image: 'mesosphere/aws-cli:latest'
```
Comment (sameersbn) [inline Jenkinsfile]: Avoid using `latest` tag

File: Jenkinsfile [hunk 9]
```diff
@@ -490,6 +506,107 @@ az aks create                      \
             }
         }
     }
+*/
+
+    //def eksVersions = ["1.10", "1.11"]
+    def eksVersions = ["1.11"]
```
Comment (sameersbn) [inline Jenkinsfile]: reminder to enable tests for other supported k8s versions

File: Jenkinsfile [hunk 10]
```diff
@@ -490,6 +506,107 @@ az aks create                      \
...
+                                            // Create EKS cluster: requires AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY environment
+                                            // variables to reference a user in AWS with the correct privileges to create an EKS
+                                            // cluster: https://github.com/weaveworks/eksctl/issues/204#issuecomment-450280786
+                                            container('eksctl') {
+                                                sh "eksctl create cluster --name ${clusterName} --version ${kversion} --node-type m5.large --nodes 3 --nodes-min 3 --nodes-max 3 --timeout 60m --kubeconfig \$KUBECONFIG --tags 'platform=${platform},branch=${BRANCH_NAME},build=${BUILD_URL}'"
+                                                withEnv(["PATH+KTOOL=${tool 'kubectl'}"]) {
+                                                    sh "kubectl create clusterrolebinding cluster-system-anonymous --clusterrole cluster-admin --user system:anonymous"
```
Comment (anguslees) [inline Jenkinsfile]: Erm .. I think this line is granting `cluster-admin` to a huge number of cluster users, which at least deserves a long explanatory comment.

Why is this necessary?  Is there anything else we can do instead?

File: Jenkinsfile (hunk 6 above)
Comment (anguslees) [reply Jenkinsfile]: Also: we should run the tests on the same side of the planet as jenkins, or all requests to/from the cluster are unnecessarily slow :stuck_out_tongue: 
Best would be us-east-something.

File: Jenkinsfile [hunk 11]
```diff
@@ -490,6 +506,107 @@ az aks create                      \
...
+                                                }
+                                                waitForRollout("kube-system", 30)
+                                            }
+
+                                            // NB: writeJSON doesn't work without approvals(?)
+                                            // See https://issues.jenkins-ci.org/browse/JENKINS-44587
```
Comment (anguslees) [inline Jenkinsfile]: Not a relevant comment here - this isn't attempting to write a JSON file.

File: Jenkinsfile [hunk 12]
```diff
@@ -490,6 +506,107 @@ az aks create                      \
...
+}
+"""
+                                            ])
+                                        }{
+                                            // update glue records in parent zone
+                                            def nameServers = []
+                                            container('aws') {
+                                                withEnv(["PATH+JQ=${tool 'jq'}"]) {
+                                                    def output = sh(returnStdout: true, script: """
+id=\$(aws route53 list-hosted-zones-by-name --dns-name "${dnsZone}" --max-items 1 | jq -r .HostedZones[0].Id)
```
Comment (anguslees) [inline Jenkinsfile]: You can drop the `jq` requirement with:

```suggestion
id=\$(aws route53 list-hosted-zones-by-name --dns-name "${dnsZone}" --max-items 1 --query 'HostedZones[0].Id' --output text)
```

File: Jenkinsfile [hunk 13]
```diff
@@ -490,6 +506,107 @@ az aks create                      \
...
+aws route53 get-hosted-zone --id "\$id" --query DelegationSet.NameServers
+"""
+                                                    )
+                                                    for (ns in readJSON(text: output)) {
```
Comment (anguslees) [inline Jenkinsfile]: Why is this a loop?  I think we can just do `nameServers = readJSON(text: output)`

File: Jenkinsfile [hunk 14]
```diff
@@ -122,6 +125,10 @@ def runIntegrationTest(String description, String kubeprodArgs, String ginkgoArg
                         junit 'junit/*.xml'
                     }
                 }
+
+                withEnv(["PATH+KUBECFG=${tool 'kubecfg'}"]) {
+                    sh "kubecfg -v delete kubeprod-manifest.jsonnet"
```
Comment (anguslees) [inline Jenkinsfile]: Is this useful, or does it just introduce delays?  We're about to delete the entire cluster anyway, it would be nice if that was sufficient..

(Or: if we _need_ to delete all the contents of the cluster first, then presumably we need to wait for the implications of that to complete - wait until PVCs are unbound and PVs are removed, loadbalancers are deleted, etc.  Just doing this delete here and then immediately continuing seems inadequate.)

File: Jenkinsfile [hunk 15]
```diff
@@ -490,6 +506,107 @@ az aks create                      \
...
+                                            sh "eksctl delete cluster --name ${clusterName} --timeout 60m"
+                                        }
+                                        container('aws') {
+                                            withEnv([
+                                                "PATH+JQ=${tool 'jq'}",
+                                                "HOME=${env.WORKSPACE}",
+                                            ]) {
+                                                sh """
+CONFIG="${clusterName}-autogen.json"
+ACCOUNT=\$(aws sts get-caller-identity | jq -r .Account)
```
Comment (anguslees) [inline Jenkinsfile]:
```suggestion
ACCOUNT=\$(aws sts get-caller-identity --query Account --output text)
```

File: Jenkinsfile [hunk 16]
```diff
@@ -490,6 +506,107 @@ az aks create                      \
...
+DNS_ZONE=\$(cat \${CONFIG} | jq -r .dnsZone)
```
Comment (anguslees) [inline Jenkinsfile]: Is this just `${dnsZone}`?  If so, we may as well just use the groovy variable directly.

File: Jenkinsfile (hunk 14 above)
Comment (falfaro) [reply Jenkinsfile]: Deleting the cluster using `eksctl` doesn't seem enough: it does not remove any ELB interfaces which prevent the CloudFormation stacks from being deleted. If we are to wait for the cluster to be deleted, how can I do that?

File: Jenkinsfile (hunk 12 above)
Comment (falfaro) [reply Jenkinsfile]: Done.

File: Jenkinsfile (hunk 11 above)
Comment (falfaro) [reply Jenkinsfile]: Done.

File: Jenkinsfile (hunk 10 above)
Comment (falfaro) [reply Jenkinsfile]: It is necessary because, without this, the tests for reasons I don't know yet, run as an anonymous user. I tested that this is the case even from my laptop. Even when `kubectl` and `kubecfg` work fine. So there must be something in the initialization code in the tests that create an anonymous Kubernetes client.

File: Jenkinsfile (hunk 15 above)
Comment (falfaro) [reply Jenkinsfile]: Done.

File: Jenkinsfile (hunk 16 above)
Comment (falfaro) [reply Jenkinsfile]: Done.

File: Jenkinsfile (hunk 8 above)
Comment (falfaro) [reply Jenkinsfile]: Done.

File: Jenkinsfile (hunk 7 above)
Comment (falfaro) [reply Jenkinsfile]: Done.

File: Jenkinsfile (hunk 9 above)
Comment (falfaro) [reply Jenkinsfile]: Yes

File: Jenkinsfile (hunk 6 above)
Comment (falfaro) [reply Jenkinsfile]: Done

File: Jenkinsfile (hunk 10 above)
Comment (falfaro) [reply Jenkinsfile]: Updating the Kubernetes Go client fixed this.

File: Jenkinsfile (hunk 5 above)
Comment (falfaro) [reply Jenkinsfile]: Done.

File: Jenkinsfile (hunk 14 above)
Comment (anguslees) [reply Jenkinsfile]: Your `waitForDestroy()` looks like a reasonable solution.

File: Jenkinsfile [hunk 17]
```diff
@@ -490,6 +532,115 @@ az aks create                      \
...
+                                        container('eksctl') {
+                                            sh "eksctl delete cluster --name ${clusterName} --timeout 60m"
+                                        }
+                                        container('aws') {
+                                            withEnv([
+                                                "PATH+JQ=${tool 'jq'}",
+                                                "HOME=${env.WORKSPACE}",
+                                            ]) {
+                                                sh """
+                                                    #!/bin/bash
```
Comment (anguslees) [inline Jenkinsfile]: Add `set -xe`  (jenkins does this automatically with the default `sh` shell, but not if you provide your own #! line)

File: Jenkinsfile [hunk 18]
```diff
@@ -490,6 +532,115 @@ az aks create                      \
...
+                                                    aws iam delete-policy --policy-arn "arn:aws:iam::\${ACCOUNT}:policy/bkpr-${dnsZone}"
+                                                    ACCESS_KEY_ID=\$(cat \${CONFIG} | jq -r .externalDns.aws_access_key_id)
+                                                    aws iam delete-access-key --user-name "bkpr-${dnsZone}" --access-key-id "\${ACCESS_KEY_ID}"
+                                                    aws iam delete-user --user-name "bkpr-${dnsZone}"
+                                                    CLIENT_ID=\$(cat \${CONFIG} | jq -r .oauthProxy.client_id)
+                                                    aws cognito-idp delete-user-pool-client --user-pool-id "${awsUserPoolId}" --client-id "\${CLIENT_ID}"
+                                                    DNS_ZONE_ID=\$(aws route53 list-hosted-zones-by-name --dns-name "${dnsZone}" --max-items 1 --query 'HostedZones[0].Id' --output text)
+                                                    aws route53 list-resource-record-sets --hosted-zone-id \${DNS_ZONE_ID} | jq -c '.ResourceRecordSets[]' | while read -r resourcerecordset; do
```
Comment (anguslees) [inline Jenkinsfile]: I haven't tested it in anger, but I think the following is the same fwiw.  (apparently `--query` can do all sorts of things!)

```sh
aws route53 list-resource-record-sets \
  --hosted-zone-id \${DNS_ZONE_ID} \
  --query '{ChangeBatch:{Changes:ResourceRecordSets[?Type != `NS` && Type != `SOA`].{Action:`DELETE`,ResourceRecordSet:@}}}' \
  --output json > changes

aws route53 change-resource-record-sets \
  --cli-input-json file://changes \
  --hosted-zone-id \${DNS_ZONE_ID} \
  --query 'ChangeInfo.Id' \
  --output text

aws route53 delete-hosted-zone \
  --id \${DNS_ZONE_ID} \
  --query 'ChangeInfo.Id' \
  --output text
```

File: Jenkinsfile (hunk 9 above)
Comment (falfaro) [reply Jenkinsfile]: Done.

File: Jenkinsfile (hunk 18 above)
Comment (falfaro) [reply Jenkinsfile]: Works!

File: Jenkinsfile (hunk 17 above)
Comment (falfaro) [reply Jenkinsfile]: Done.

**Comment by falfaro** [comment]
PTAL

File: Jenkinsfile (hunk 13 above)
Comment (anguslees) [reply Jenkinsfile]: Oops, sorry.  Looks like `readJSON` returns a `JSONArray`, but the `insertGlueRecords` function requires an `ArrayList`.  Need to put the loop back (boo!), or change `insertGlueRecords` to take a `java.util.List` (yay!) instead.

File: Jenkinsfile (hunk 9 above)
Comment (sameersbn) [reply Jenkinsfile]: i guess reverted :slightly_smiling_face:

File: Jenkinsfile [hunk 19]
```diff
@@ -490,6 +499,150 @@ az aks create                      \
...
+                                        if (fileExists("${env.WORKSPACE}/.kubeconf")) {
+                                            // Uninstall BKPR
+                                            withEnv([
+                                                "PATH+KUBECFG=${tool 'kubecfg'}",
+                                                "PATH+KUBECTL=${tool 'kubectl'}",
+                                                "KUBECONFIG=${env.WORKSPACE}/.kubeconf",
+                                            ]) {
+                                                // This is required for AWS in order to release/destroy ELB network interfaces.
+                                                sh """
+                                                    #!/bin/bash
```
Comment (sameersbn) [inline Jenkinsfile]: is this shebang really needed? and also the following `set` commands. I believe `set -x` is the default

File: Jenkinsfile [hunk 20]
```diff
@@ -85,7 +86,7 @@ String gcpLabel(String s) {
 }
 
 def runIntegrationTest(String description, String kubeprodArgs, String ginkgoArgs, Closure clusterSetup, Closure dnsSetup) {
-    timeout(120) {
+    timeout(240) {
```
Comment (sameersbn) [inline Jenkinsfile]: just noticed this, can it really take 4 hours for the tests to complete. Is this a realistic number?

File: Jenkinsfile [hunk 21]
```diff
@@ -84,45 +85,48 @@ String gcpLabel(String s) {
...
+                } catch (error) {
+                    sh "kubectl --namespace kubeprod get po,deploy,svc,ing"
+                    input 'Paused for manual debugging'
+                    throw error
+                } finally {
+                    junit 'junit/*.xml'
                 }
             }
+
+            clusterDestroy()
```
Comment (anguslees) [inline Jenkinsfile]: With this change, `clusterDestroy()` will not be run when any of the previous steps fail (raise an exception).  I think we need to run `clusterDestroy()` in some sort of try/finally block (see the previous code structure).

File: Jenkinsfile (hunk 19 above)
Comment (anguslees) [reply Jenkinsfile]: It is not needed here (afaics), and we should revert to standard `sh` (without the shebang, and without explicit `set +x`/`set -x` (?!))

File: Jenkinsfile [hunk 22]
```diff
@@ -490,6 +502,152 @@ az aks create                      \
...
+                                                """
+
+                                                // Destroy AWS objects
+                                                container('aws') {
+                                                    withEnv([
+                                                        "PATH+JQ=${tool 'jq'}",
+                                                        "HOME=${env.WORKSPACE}",
+                                                    ]) {
+                                                        sh """
+                                                            #!/bin/bash
```
Comment (anguslees) [inline Jenkinsfile]: Ditto, I don't believe this contains any bash-specific constructs anymore (an earlier version used `<<<`)

Comment (falfaro) [reply Jenkinsfile]: Fixed.

File: Jenkinsfile (hunk 20 above)
Comment (falfaro) [reply Jenkinsfile]: Reverted to 120.

File: Jenkinsfile (hunk 21 above)
Comment (falfaro) [reply Jenkinsfile]: Wrapped inside a finally {}.

File: Jenkinsfile (hunk 19 above)
Comment (falfaro) [reply Jenkinsfile]: Fixed.

File: Jenkinsfile [hunk 23]
```diff
@@ -255,7 +270,7 @@ spec:
         }, failFast: true)
 
     def platforms = [:]
-
+/*
```
Comment (anguslees) [inline Jenkinsfile]: Still need to restore testing on other platforms.  You may need to change their cleanup code to follow your `try/finally` refactor.

Comment (falfaro) [reply Jenkinsfile]: I'm on it :-)

File: Jenkinsfile (hunk 13 above)
Comment (falfaro) [reply Jenkinsfile]: Fixed.

**Comment by falfaro** [comment]
@anguslees @sameersbn PTAL

File: tests/vendor/vendor.json [hunk 24]
```diff
@@ -2,6 +2,10 @@
 	"comment": "",
 	"ignore": "test appengine",
 	"package": [
+		{
```
Comment (anguslees) [inline tests/vendor/vendor.json]: This looks like the result of an incorrect govendor invocation.  We should remove this.

File: tests/vendor/vendor.json [hunk 25]
```diff
@@ -44,6 +48,11 @@
 			"revision": "de5bf2ad457846296e2031421a34e2568e304e35",
 			"revisionTime": "2017-08-10T14:37:23Z"
 		},
+		{
```
Comment (anguslees) [inline tests/vendor/vendor.json]: ditto. We should remove this.

Comment (falfaro) [reply tests/vendor/vendor.json]: Fixed.

File: tests/vendor/vendor.json (hunk 24 above)
Comment (falfaro) [reply tests/vendor/vendor.json]: Fixed.

**Comment by falfaro** [comment]
@anguslees @sameersbn please take another look as I think this is now robust enough to be merged.

**Comment by sameersbn** [comment]
bors retry

**Comment by sameersbn** [comment]
bors retry

File: Jenkinsfile [hunk 26]
```diff
@@ -111,7 +111,9 @@ def runIntegrationTest(String description, String kubeprodArgs, String ginkgoArg
                     sh "kubectl --namespace kube-system get po,deploy,svc,ing"
 
                     try {
-                        sh "./bin/kubeprod -v=1 install ${kubeprodArgs} --manifests=manifests"
+                        retry(3) {
```
Comment (anguslees) [inline Jenkinsfile]: Instead, is it sufficient to run `kubectl api-versions` somewhere above - and loop/delay until that succeeds?  Hopefully that will delay long enough that whatever `kubeprod install` is hitting will have stabilised.  Again, I would rather we didn't treat `kubeprod install` specially at all, because flakiness in this command is flakiness that the user will see (and we should instead fix `kubeprod install`).

Comment (sameersbn) [reply Jenkinsfile]: hmm.. that makes sense. i'll make this change
